File: apps/mobile2/compose/src/commonTest/kotlin/co/typie/screen/more/stats/StatsCalculatorTest.kt
package co.typie.screen.more.stats

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlinx.datetime.LocalDate

class StatsCalculatorTest {

  @Test
  fun calculateStreakData_countsCurrentLongestAndMonthlyDays() {
    val changes =
      listOf(
        StatsCharacterCountChange(date = LocalDate(2026, 3, 17), additions = 40, deletions = 0),
        StatsCharacterCountChange(date = LocalDate(2026, 3, 19), additions = 80, deletions = -5),
        StatsCharacterCountChange(date = LocalDate(2026, 3, 20), additions = 120, deletions = -8),
        StatsCharacterCountChange(date = LocalDate(2026, 3, 21), additions = 160, deletions = -13),
      )

    val result =
      calculateStreakData(
        characterCountChanges = changes,
        totalCharacterCount = 1_000,
        today = LocalDate(2026, 3, 21),
      )

    assertEquals(
      ActivityData(
        currentStreak = 3,
        longestStreak = 3,
        thisMonthActiveDays = 4,
        totalActiveDays = 4,
        averageCharacterCountPerDay = 250,
      ),
      result,
    )
  }

  @Test
  fun calculateStreakData_startsFromYesterdayWhenTodayHasNoActivity() {
    val changes =
      listOf(
        StatsCharacterCountChange(date = LocalDate(2026, 3, 19), additions = 70, deletions = 0),
        StatsCharacterCountChange(date = LocalDate(2026, 3, 20), additions = 95, deletions = 0),
      )

    val result =
      calculateStreakData(
        characterCountChanges = changes,
        totalCharacterCount = 500,
        today = LocalDate(2026, 3, 21),
      )

    assertEquals(2, result.currentStreak)
    assertEquals(2, result.longestStreak)
    assertEquals(2, result.thisMonthDays)
    assertEquals(2, result.totalDays)
    assertEquals(250, result.avgCharactersPerDay)
  }

  @Test
  fun calculateWeekdayPattern_groupsBySundayFirstAndRoundsAverage() {
    val changes =
      listOf(
        StatsCharacterCountChange(date = LocalDate(2026, 3, 22), additions = 120, deletions = 0),
        StatsCharacterCountChange(date = LocalDate(2026, 3, 23), additions = 80, deletions = 0),
        StatsCharacterCountChange(date = LocalDate(2026, 3, 23), additions = 20, deletions = 0),
        StatsCharacterCountChange(date = LocalDate(2026, 3, 24), additions = 0, deletions = -10),
      )

    val result = calculateWeekdayPattern(changes)

    assertEquals(
      listOf(
        WeekdayActivityData(
          dayIndex = 0,
          label = "일",
          totalAdditions = 120,
          averageAdditions = 120,
          activeDays = 1,
        ),
        WeekdayActivityData(
          dayIndex = 1,
          label = "월",
          totalAdditions = 100,
          averageAdditions = 50,
          activeDays = 2,
        ),
        WeekdayActivityData(
          dayIndex = 2,
          label = "화",
          totalAdditions = 0,
          averageAdditions = 0,
          activeDays = 0,
        ),
        WeekdayActivityData(
          dayIndex = 3,
          label = "수",
          totalAdditions = 0,
          averageAdditions = 0,
          activeDays = 0,
        ),
        WeekdayActivityData(
          dayIndex = 4,
          label = "목",
          totalAdditions = 0,
          averageAdditions = 0,
          activeDays = 0,
        ),
        WeekdayActivityData(
          dayIndex = 5,
          label = "금",
          totalAdditions = 0,
          averageAdditions = 0,
          activeDays = 0,
        ),
        WeekdayActivityData(
          dayIndex = 6,
          label = "토",
          totalAdditions = 0,
          averageAdditions = 0,
          activeDays = 0,
        ),
      ),
      result,
    )
  }

  @Test
  fun generateActivityChartDays_fillsMissingDatesAndNormalizesDeletions() {
    val result =
      generateActivityChartDays(
        characterCountChanges =
          listOf(
            StatsCharacterCountChange(
              date = LocalDate(2026, 3, 20),
              additions = 60,
              deletions = -20,
            ),
            StatsCharacterCountChange(date = LocalDate(2026, 3, 22), additions = 40, deletions = 0),
          ),
        endDate = LocalDate(2026, 3, 22),
        dayCount = 3,
      )

    assertEquals(
      listOf(
        StatsActivityDay(date = LocalDate(2026, 3, 20), additions = 60, deletions = 20),
        StatsActivityDay(date = LocalDate(2026, 3, 21), additions = 0, deletions = 0),
        StatsActivityDay(date = LocalDate(2026, 3, 22), additions = 40, deletions = 0),
      ),
      result,
    )
  }
}

File: legacy/mobile/test/screens/stats/stats_calculator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:jiffy/jiffy.dart';
import 'package:typie/screens/stats/stats_calculator.dart';

void main() {
  test('calculateStreakData computes streaks and monthly counts', () {
    final changes = [
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-06'), additions: 120, deletions: 5),
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-05'), additions: 80, deletions: 3),
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-03'), additions: 40, deletions: 1),
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-01'), additions: 20, deletions: 0),
      StatsCharacterCountChange(date: Jiffy.parse('2026-02-28'), additions: 10, deletions: 0),
    ];

    final data = calculateStreakData(changes, 1000, now: Jiffy.parse('2026-03-06'));

    expect(data.currentStreak, 2);
    expect(data.longestStreak, 2);
    expect(data.thisMonthDays, 4);
    expect(data.totalDays, 5);
    expect(data.avgCharactersPerDay, 200);
  });

  test('calculateStreakData starts from yesterday when today is inactive', () {
    final changes = [
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-05'), additions: 90, deletions: 0),
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-04'), additions: 70, deletions: 0),
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-02'), additions: 10, deletions: 0),
    ];

    final data = calculateStreakData(changes, 300, now: Jiffy.parse('2026-03-06'));

    expect(data.currentStreak, 2);
    expect(data.longestStreak, 2);
    expect(data.thisMonthDays, 3);
    expect(data.totalDays, 3);
    expect(data.avgCharactersPerDay, 100);
  });

  test('calculateWeekdayPattern returns sunday-based indices and averages', () {
    final changes = [
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-01'), additions: 10, deletions: 0), // 일
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-08'), additions: 30, deletions: 0), // 일
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-02'), additions: 20, deletions: 0), // 월
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-09'), additions: 40, deletions: 0), // 월
      StatsCharacterCountChange(date: Jiffy.parse('2026-03-11'), additions: 0, deletions: 0),
    ];

    final data = calculateWeekdayPattern(changes);

    expect(data, hasLength(7));
    expect(data[0].label, '일');
    expect(data[0].count, 2);
    expect(data[0].avgAdditions, 20);
    expect(data[1].label, '월');
    expect(data[1].count, 2);
    expect(data[1].avgAdditions, 30);
    expect(data[3].count, 0);
    expect(data[3].avgAdditions, 0);
  });
}
